test/callback.tests.js

```diff
@@ -0,0 +1,327 @@
+        _state: '__test_state__'
+      },
+      body: {
+        state: '__test_state__',
+        id_token: '__invalid_token__'
+      }
+    });
+    assert.equal(statusCode, 400);
+    assert.equal(err.message, 'checks.state argument is missing');
+  });
+
+  it('should not strip claims when using custom claim filtering', async () => {
+    const { currentUser } = await setup({
+      authOpts: {
+        identityClaimFilter: []
+      },
+      cookies: {
+        _state: expectedDefaultState,
+        _nonce: '__test_nonce__'
+      },
+      body: {
+        state: expectedDefaultState,
+        id_token: makeIdToken()
+      }
+    });
+    assert.equal(currentUser.iss, 'https://op.example.com/');
+    assert.equal(currentUser.aud, clientID);
+    assert.equal(currentUser.nonce, '__test_nonce__');
+    assert.exists(currentUser.iat);
+    assert.exists(currentUser.exp);
+  });
+
+  it('should expose the id token when id_token is valid', async () => {
+    const idToken = makeIdToken();
+    const { response: { statusCode, headers }, currentUser, tokens } = await setup({
+      cookies: {
+        _state: expectedDefaultState,
+        _nonce: '__test_nonce__'
+      },
+      body: {
+        state: expectedDefaultState,
+        id_token: idToken
+      }
+    });
+    assert.equal(statusCode, 302);
+    assert.equal(headers.location, 'https://example.org');
+    assert.ok(currentUser);
+    assert.equal(currentUser.sub, '__test_sub__');
+    assert.equal(currentUser.nickname, '__test_nickname__');
+    assert.notExists(currentUser.iat);
+    assert.notExists(currentUser.iss);
+    assert.notExists(currentUser.aud);
+    assert.notExists(currentUser.exp);
+    assert.notExists(currentUser.nonce);
+    assert.equal(tokens.isAuthenticated, true);
+    assert.equal(tokens.idToken, idToken);
+    assert.isUndefined(tokens.refreshToken);
+    assert.isUndefined(tokens.accessToken);
+    assert.include(tokens.idTokenClaims, {
+      sub: '__test_sub__'
+    });
+  });
+
+  it('should expose all tokens when id_token is valid and response_type is \'code id_token\'', async () => {
+    const idToken = makeIdToken({
+      c_hash: '77QmUPtjPfzWtF2AnpK9RQ'
+    });
+
+    const { tokens } = await setup({
+      authOpts: {
+        clientSecret: '__test_client_secret__',
+        authorizationParams: {
+          response_type: 'code id_token',
+          audience: 'https://api.example.com/',
+          scope: 'openid profile email read:reports offline_access'
+        }
+      },
+      cookies: {
+        _state: expectedDefaultState,
+        _nonce: '__test_nonce__'
+      },
+      body: {
+        state: expectedDefaultState,
+        id_token: idToken,
+        code: 'jHkWEdUXMU1BwAsC4vtUsZwnNvTIxEl0z9K3vx5KF0Y',
+      }
+    });
+
+    assert.equal(tokens.isAuthenticated, true);
+    assert.equal(tokens.idToken, idToken);
+    assert.equal(tokens.refreshToken, '__test_refresh_token__');
+    assert.include(tokens.accessToken, {
```

Comment:
we ought to test also that

- `tokens.accessToken.expires_in` being a number (its required to be returned by all access token responses) that counts down every second
- `tokens.accessToken.isExpired` returns true/false

Reply:
0c6814b